package sae;
// Options
option py_generic_services = true;

//
// RPC Transport
//
message Message {
    required uint32   id       = 1; // Transaction id
    optional Request  request  = 2;
    optional Response response = 3;
    optional Error    error    = 4;
}

message Request {
    required string method             = 1;
    required bytes  serialized_request = 2; // encoded *Request
}

message Response {
    optional bytes  serialized_response = 1; // encoded *Response
}

message Error {
    required ErrorCode code = 1;
    optional string    text = 2;
}
//
//
//
enum ErrorCode {
    ERR_OK                  = 0; //
    ERR_INTERNAL            = 1; // Unhandled exception on server side during request processing
    ERR_INVALID_METHOD      = 2; // Invalid method name
    ERR_INVALID_TRANSACTION = 3; // Transaction does not exists
    ERR_TRANSACTION_EXISTS  = 4;
    ERR_UNKNOWN_ACTIVATOR   = 5;
    ERR_INVALID_PROFILE     = 6; // Invalid profile
    ERR_INVALID_SCHEME      = 7; // Invalid access scheme
    ERR_UNKNOWN_EVENT_SOURCE = 8;
    ERR_AUTH_FAILED         = 9; // Authentication failed
    ERR_AUTH_REQUIRED       =10; // Authentication required
    ERR_INVALID_UPGRADE     =11; // Invalid file requested for upgrade
    ERR_OVERLOAD            =12; // System limit exceeded. Retry operation after delay
    ERR_INVALID_SCRIPT      =13; // Invalid script
    ERR_SCRIPT_EXCEPTION    =14; // Script terminated with exception
    ERR_ACTIVATOR_NOT_AVAILABLE =15; // Activator not available now
    ERR_DOWN                =16; // Host is down
    ERR_TIMEOUT             =17; // Script timed out
    ERR_ACTIVATOR_LOST      =18; // Connection to activator lost during transaction
    ERR_LOGIN_FAILED        =19; // Failed to log in
    ERR_CANCELLED           =20; // Script cancelled
    ERR_NOT_SUPPORTED       =21; // Not supported feature
    ERR_HTTP_ERROR          =22; // HTTP Error
    ERR_INVALID_SHARD       =23; // Shard mismatch
    ERR_SHARD_IS_DOWN       =24; // Shard is disabled
    ERR_OBJ_OVERLOAD        =25; // Too many scripts on object
    ERR_PROTO_MISMATCH      =26; // Protocol version mismatch
    ERR_SETUP_FAILED        =27; // Cannot negotiate crypto session
    ERR_OBJECT_NOT_MANAGED  =28; // Object is not managed
}

//
//
//
enum AccessScheme {
    TELNET = 0;
    SSH    = 1;
    HTTP   = 2;
}

message AccessProfile {
    message KWArg {
        required string key = 1;
        required string value = 2;
    }
    message Caps {
        required string   capability = 1;
        optional string   str_value = 2;
        optional sfixed64 int_value = 3;
        optional bool     bool_value = 4;
        optional float    float_value = 5;
    }
    required string       profile        = 1;
    required AccessScheme scheme         = 2;
    required string       address        = 3;
    optional uint32       port           = 4;
    optional string       user           = 5;
    optional string       password       = 6;
    optional string       super_password = 7;
    optional string       path           = 8;
    optional string       snmp_ro        = 9;
    optional string       snmp_rw        =10;
    repeated KWArg        attrs          =11;
    repeated Caps         caps           =12;
}
//
// Protocol negotiation
// Activator -> SAE
//
message ProtocolRequest {
    required string protocol = 1;
    required string version = 2;
}

message ProtocolResponse {
    required string protocol = 1;
    required string version = 2;
}

//
// Crypto algorithm negotiation
//
message SetupRequest {
    repeated string key_exchanges = 1;
    repeated string public_keys = 2;
    repeated string ciphers = 3;
    repeated string macs = 4;
    repeated string compressions = 5;
}

message SetupResponse {
    required string key_exchange = 1;
    required string public_key = 2;
    required string cipher = 3;
    required string mac = 4;
    required string compression = 5;
}

//
// Key exhange
//
message KEXRequest {
    message DHGroup1Sha1 {
        required bytes e = 1;
    }
    
    optional DHGroup1Sha1 dh_group1_sha1 = 1;
}

message KEXResponse {
    message DHGroup1Sha1 {
        required bytes e = 1;
    }
    
    optional DHGroup1Sha1 dh_group1_sha1 = 1;
}

//
// Ping server
// SAE->Activator
// Activator->SAE
message PingRequest {
}

message PingResponse {
}
//
// Register activator at the SAE
// Activator->SAE
message RegisterRequest {
    required string name = 1;
}

message RegisterResponse {
    required string nonce = 1;
}
//
//
//
message AuthRequest {
    required string name   = 1;
    required string digest = 2;
    required int32 max_scripts = 3;
    required string instance = 4;
}

message AuthResponse {
}

//
// Execute script
//
message ScriptRequest {
    message KWArg {
        required string key = 1;
        required string value = 2;
    }
    required AccessProfile access_profile = 1;
    required string        object_name = 2;
    required string        script  = 3;
    repeated KWArg         kwargs  = 4;
    optional uint32        timeout = 5;
}

message ScriptResponse {
    required string result = 1;
}

//
// Ping hosts
//
message ObjectStatusRequest {
    message ObjectStatusItem {
        required string object = 1;
        required bool status = 2;
    }
    repeated ObjectStatusItem status = 1;
}

message ObjectStatusResponse {
}

message PingCheckRequest {
    repeated string addresses = 1;
}

message PingCheckResponse {
    message PingStatusItem {
        required string address = 1;
        required bool status = 2;
    }
    repeated PingStatusItem status = 1;
}

message StatusRequest {}

message StatusResponse {
    message ScriptInfo {
        required string script = 1;
        required string address = 2;
        required int32 start_time = 3;
        required int32 timeout = 4;
        required string object_name = 5;
    }
    required int32 timestamp = 1;
    required string pool = 2;
    required string instance = 3;
    required string state = 4;
    required int32 last_state_change = 5;
    required int32 max_scripts = 6;
    required int32 current_scripts = 7;
    required int32 scripts_processed = 8;
    required int32 scripts_failed = 9;
    repeated ScriptInfo scripts = 10;
}
//
// RPC Services
//
service SAEService {
    rpc protocol         (ProtocolRequest)        returns (ProtocolResponse);
    rpc setup            (SetupRequest)           returns (SetupResponse);
    rpc kex              (KEXRequest)             returns (KEXRequest);
    rpc ping             (PingRequest)            returns (PingResponse);
    rpc register         (RegisterRequest)        returns (RegisterResponse);
    rpc auth             (AuthRequest)            returns (AuthResponse);
    rpc script           (ScriptRequest)          returns (ScriptResponse);
    rpc object_status    (ObjectStatusRequest)    returns (ObjectStatusResponse);
    rpc get_status       (StatusRequest)          returns (StatusResponse);
    rpc ping_check       (PingCheckRequest)       returns (PingCheckResponse);
}
